### PR TITLE
fix layout for iframe embed

### PR DESCRIPTION
**Pull request by penk85**
WPC-572: Fix sizing of iframe embeds. 
[Editor](https://editor.bwpc572sp.wepublish.dev/)
[Website](https://bwpc572sp.wepublish.dev/a/zk4jiLO6G6NFbU0T/spotify-embeds)

WhatsApp Image 2022-03-29 at 20 17 39
WhatsApp Image 2022-03-29 at 20 18 47

### CODE REVIEW

**Review by camilleory — APPROVED** [review]

File: examples/website/src/shared/utility.ts [hunk 1]
```diff
@@ -125,6 +125,16 @@ export function transformCssStringToObject(styleCustom: string): object {
...
+      const index = key.indexOf('-')
+      if (index !== -1) {
+        let camelCaseKey = key.substring(index)
+        camelCaseKey = camelCaseKey.slice(1).replace(camelCaseKey[1], camelCaseKey[1].toUpperCase())
+
+        return Object.assign(previousValue, {
+          [(key.substring(0, index) + camelCaseKey).trim()]: value.trim()
+        })
+      }
+
```
Comment (tomaszdurka) [inline examples/website/src/shared/utility.ts:137]: This seems to cover only single dash cases - I think there might be more.

Wdyt about using something more generic for the `key`?
```
key.toLowerCase().replace(/\-(.)/gm, ($0, $1) => $1.toUpperCase())
```

**Review by tomaszdurka — CHANGES_REQUESTED** [review]

File: examples/website/src/shared/utility.ts (hunk 1 above)
Comment (penk85) [reply examples/website/src/shared/utility.ts:137]: @tomaszdurka I'm not sure I understand the second half of this, but it indeed works.

File: examples/website/src/shared/blocks/embedBlock.tsx [hunk 2]
```diff
@@ -100,7 +100,7 @@ function embedForData(data: EmbedData) {
             allowFullScreen
             title={data.title}
             src={data.url}
-            width={data.width}
+            width="100%"
```
Comment (michael-scheurer) [inline examples/website/src/shared/blocks/embedBlock.tsx:103]: In my opinion it should overtake data.width per default and if data.width is not set a.k. undefined use 100%
Like so:
`width={data.wdith || "100%"}`

File: examples/website/src/shared/utility.ts [hunk 3]
```diff
@@ -125,7 +125,12 @@ export function transformCssStringToObject(styleCustom: string): object {
   return styleRules.reduce((previousValue: object, currentValue: string) => {
     const [key, value] = currentValue.split(':')
     if (key && value) {
-      return Object.assign(previousValue, {[key.trim()]: value.trim()})
+      return Object.assign(previousValue, {
```
Comment (michael-scheurer) [inline examples/website/src/shared/utility.ts:128]: what function! don't get, why the heck it is necessary to write such functions. but if it works still for other iframes, it's ok for me.

File: packages/editor/src/client/blocks/embeds/iframe.tsx [hunk 4]
```diff
@@ -41,7 +40,9 @@ export function IframeEmbed({url, title, width, height, styleCustom, sandbox}: I
           scrolling="no"
           frameBorder="0"
           allowFullScreen
-          sandbox={sandbox}
+          width="100%"
+          height={height}
+          sandbox={sandbox?.length !== 0 ? sandbox : undefined}
```
Comment (michael-scheurer) [inline packages/editor/src/client/blocks/embeds/iframe.tsx:45]: U can write this shorter, because 0 is a falsy value.

`sandbox={!!sandbox?.length ? sandbox : undefined}`

**Review by michael-scheurer — DISMISSED** [review]

File: examples/website/src/shared/blocks/embedBlock.tsx (hunk 2 above)
Comment (penk85) [reply examples/website/src/shared/blocks/embedBlock.tsx:103]: Actually you are right it doesn't make sense to override the data. There was an underlying issue here which was causing issues with the width -it was saved as number format and therefore lost the '%' sign when rendering. Iframe with 100% width got saved as 100px width. I restored the original `data.width` here and it should be working.

**Review by camilleory — DISMISSED** [review]

**Review by tomaszdurka — APPROVED** [review]

File: packages/editor/src/client/api/index.ts [hunk 5]
```diff
@@ -342,17 +342,17 @@ export type EmbedBlock = {
...
   styleCustom?: Maybe<Scalars['String']>;
   sandbox?: Maybe<Scalars['String']>;
 };
 
 export type EmbedBlockInput = {
   url?: Maybe<Scalars['String']>;
   title?: Maybe<Scalars['String']>;
-  width?: Maybe<Scalars['Int']>;
-  height?: Maybe<Scalars['Int']>;
+  width?: Maybe<Scalars['String']>;
```
Comment (elias-summermatter) [inline packages/editor/src/client/api/index.ts:354]: @tomaszdurka shouldn't we add here a mongodb migration that changes int to string? Or isn't this a problem if you have mixed types? I don't know mongodb enough.

File: packages/editor/src/client/api/index.ts [hunk 6]
```diff
@@ -646,7 +646,7 @@ export type MemberPlan = {
   slug: Scalars['String'];
   image?: Maybe<Image>;
   description?: Maybe<Scalars['RichText']>;
-  tags: Array<Scalars['String']>
+  tags?: Maybe<Array<Scalars['String']>>;
```
Comment (elias-summermatter) [inline packages/editor/src/client/api/index.ts:649]: Is this change really necessary and when yes why?

File: packages/editor/src/client/blocks/embeds/iframe.tsx [hunk 7]
```diff
@@ -4,15 +4,15 @@ import {transformCssStringToObject} from '../../utility'
...
+  width?: string
+  height?: string
   styleCustom?: string
   sandbox?: string
 }
 
 export function IframeEmbed({url, title, width, height, styleCustom, sandbox}: IframeEmbedProps) {
-  const ratio = width !== undefined && height !== undefined ? width / height : 0
-  const noRatio = !!styleCustom && ratio === 0
+  const ratio = width !== undefined && height !== undefined ? parseInt(width) / parseInt(height) : 0
```
Comment (elias-summermatter) [inline packages/editor/src/client/blocks/embeds/iframe.tsx:14]: Since with and height are strings could you not write !width and !height or is there an other action if the values are null?

File: packages/editor/src/client/panel/memberPlanEditPanel.tsx [hunk 8]
```diff
@@ -108,7 +108,7 @@ export function MemberPlanEditPanel({id, onClose, onSave}: MemberPlanEditPanelPr
     if (data?.memberPlan) {
       setName(data.memberPlan.name)
       setSlug(data.memberPlan.slug)
-      setTags(data.memberPlan.tags)
+      setTags(data.memberPlan.tags ?? [])
```
Comment (elias-summermatter) [inline packages/editor/src/client/panel/memberPlanEditPanel.tsx:111]: Why this change?

File: packages/editor/src/client/utility.ts [hunk 9]
```diff
@@ -125,7 +125,12 @@ export function transformCssStringToObject(styleCustom: string): Record<string,
   return styleRules.reduce((previousValue: Record<string, unknown>, currentValue: string) => {
     const [key, value] = currentValue.split(':')
     if (key && value) {
-      return Object.assign(previousValue, {[key.trim()]: value.trim()})
+      return Object.assign(previousValue, {
+        [key
+          .toLowerCase()
+          .replace(/-(.)/gm, ($0, $1) => $1.toUpperCase())
+          .trim()]: value.trim()
+      })
```
Comment (elias-summermatter) [inline packages/editor/src/client/utility.ts:133]: Please add a comment and describe what this function does. Nobody understands this.

File: examples/website/src/shared/utility.ts [hunk 10]
```diff
@@ -125,7 +125,12 @@ export function transformCssStringToObject(styleCustom: string): object {
...
+        [key
+          .toLowerCase()
+          .replace(/-(.)/gm, ($0, $1) => $1.toUpperCase())
+          .trim()]: value.trim()
+      })
```
Comment (elias-summermatter) [inline examples/website/src/shared/utility.ts:133]: Please add a comment and describe what this function does. Nobody understands this.

**Review by elias-summermatter — CHANGES_REQUESTED** [review]

File: packages/editor/src/client/api/index.ts (hunk 5 above)
Comment (penk85) [reply packages/editor/src/client/api/index.ts:354]: I don't think this needs a migration. @tomaszdurka what do you think?

File: packages/editor/src/client/api/index.ts (hunk 6 above)
Comment (penk85) [reply packages/editor/src/client/api/index.ts:649]: It was causing errors compiling . I think others have run into the same, but should probably be fixed somewhere else.

File: packages/editor/src/client/blocks/embeds/iframe.tsx (hunk 7 above)
Comment (penk85) [reply packages/editor/src/client/blocks/embeds/iframe.tsx:14]: I assume you're right. Will change this.

File: packages/editor/src/client/utility.ts (hunk 9 above)
Comment (penk85) [reply packages/editor/src/client/utility.ts:133]: Its to extract the html properties of the pasted embed function and format as React. The function was already here, but needed to be adjusted to handle multi-word properties. `border-radius` -> `borderRadius`

Comment (penk85) [reply packages/editor/src/client/utility.ts:133]: Will add comment.

File: examples/website/src/shared/utility.ts (hunk 3 above)
Comment (penk85) [reply examples/website/src/shared/utility.ts:128]: Yes was necessary to handle properties from the iframe

File: packages/editor/src/client/panel/memberPlanEditPanel.tsx (hunk 8 above)
Comment (penk85) [reply packages/editor/src/client/panel/memberPlanEditPanel.tsx:111]: See comment above.

**Review by tomaszdurka — DISMISSED** [review]

File: packages/editor/src/client/api/index.ts (hunk 5 above)
Comment (tomaszdurka) [reply packages/editor/src/client/api/index.ts:354]: Ah, well it will come as Int, even though it should be String. It won't fail unless some string specific operations are made here.
I agree migration would help.

@penk85 can you write that quickly? @elias-summermatter can you help therE?

Comment (elias-summermatter) [reply packages/editor/src/client/api/index.ts:354]: @penk85 where you able to wrtie this migration?

File: packages/api-db-mongodb/src/migration.ts [hunk 11]
```diff
@@ -702,6 +709,73 @@ export const Migrations: Migration[] = [
...
+            if (block.type === BlockType.Embed) {
+              block.height = String(block.height)
+              block.width = String(block.width)
+            }
+          })
+        }
+        await articles.findOneAndReplace({_id: article._id}, article)
+      }
+
+      const pages = db.collection(CollectionName.Articles)
```
Comment (tomaszdurka) [inline packages/api-db-mongodb/src/migration.ts:748]: This should be `CollectionName.Pages` right?

**Review by tomaszdurka — CHANGES_REQUESTED** [review]

File: packages/api-db-mongodb/src/migration.ts (hunk 11 above)
Comment (penk85) [reply packages/api-db-mongodb/src/migration.ts:748]: Whooops yes :+1:

**Review by tomaszdurka — APPROVED** [review]
I see you found some more

**Review by tomaszdurka — DISMISSED** [review]

**Review by Itrulia — APPROVED** [review]

**Review by camilleory — APPROVED** [review]
👍

File: packages/api-db-mongodb/src/migration.ts [hunk 12]
```diff
@@ -789,6 +797,73 @@ export const Migrations: Migration[] = [
...
+  },
+  {
+    // change embed block properties width and height from number to string
+    version: 20,
+    async migrate(db) {
+      const articles = db.collection(CollectionName.Articles)
+      const migrationArticles = await articles.find().toArray()
+
+      for (const article of migrationArticles) {
+        if (article.draft) {
```
Comment (elias-summermatter) [inline packages/api-db-mongodb/src/migration.ts:809]: Why not using the mongo internal function  "$toString" insted of iterating over every article?

https://www.mongodb.com/docs/manual/reference/operator/aggregation/toString/

File: packages/api-db-mongodb/src/migration.ts [hunk 13]
```diff
@@ -789,6 +797,73 @@ export const Migrations: Migration[] = [
...
+          article.draft.blocks.forEach((block: ArticleBlock) => {
```
Comment (elias-summermatter) [inline packages/api-db-mongodb/src/migration.ts:810]: Why not use, you can prevent code duplication like this
```
for (const key in article) {
    const blocks = article[key].block
   ....
}
```
Or are there articles of a state you dont want to change?

File: packages/api-db-mongodb/src/migration.ts [hunk 14]
```diff
@@ -789,6 +797,73 @@ export const Migrations: Migration[] = [
...
+        if (page.draft) {
+          page.draft.blocks.forEach((block: PageBlock) => {
+            if (block.type === BlockType.Embed) {
+              block.height = String(block.height)
+              block.width = String(block.width)
+            }
+          })
+        }
+        if (page.published) {
+          page.published.blocks.forEach((block: PageBlock) => {
```
Comment (elias-summermatter) [inline packages/api-db-mongodb/src/migration.ts:849]: Same here

**Review by elias-summermatter — CHANGES_REQUESTED** [review]

File: packages/api-db-mongodb/src/migration.ts (hunk 12 above)
Comment (elias-summermatter) [reply packages/api-db-mongodb/src/migration.ts:809]: @tomaszdurka: @penk85  told me you told her to do it the "unsave" way. Don't you think we should use atomic operations for data manipulation that there is less risk of failing operation?

Comment (elias-summermatter) [reply packages/api-db-mongodb/src/migration.ts:809]: Or maybe use Transactions: https://www.digitalocean.com/community/tutorials/how-to-use-transactions-in-mongodb

Comment (penk85) [reply packages/api-db-mongodb/src/migration.ts:809]: @elias-summermatter transactions for migrations will be introduced in a different pull request. If that's okay, can you approve?

Comment (Itrulia) [reply packages/api-db-mongodb/src/migration.ts:809]: Unfortunately transactions in Mongodb require a replica set, we currently are only running a standalone server.

**Review by tomaszdurka — APPROVED** [review]

**Review by elias-summermatter — APPROVED** [review]
Since its approved by @tomaszdurka its fine for me.
